### PR TITLE
Virtual DOM API implementation

### LINKED ISSUE #5: Virtual DOM API Proposal
**TL;DR:** The RFC has been accomplished. **Flatten Attributes and Outputs With Nested Optional Properties and Hooks** API has been accepted and I'm gonna to implement it. Special thanks to @jedimahdi.

## RFC: Virtual DOM API

Meanwhile I'm working on Monarch elements, I'm thinking to what is the best option for virtual DOM API?!
In the following I propose multiple ways to represent virtual DOM in PureScript for Monarch. Please let me know your opinions about each one and which one you mostly prefer.

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook: { pre: const Pre }
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't separate type of the data and it can lead to name clashes as there are lots of properties and attributes with similar names in the HTML specification (property, attribute, output (event), hook, etc...)
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hookPre: const Pre
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Boilerplate
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { attrs:   mkAttrs  { href: "/" }
         , outputs: mkOutput { click: const Click
                             , blur: const Blur
                             }
         , hooks:   mkHooks  { pre: const Pre }
         }
         [ Sphoenoid.slider { attrs:   mkAttrs   { step: 1 }
                            , outputs: mkOutputs { next: const Next }
                            }
                            [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur:  const Blur
         , hooks:   mkHooks  { pre: const Pre }
         }
         [ Sphoenoid.slider { step: 1 
                            , onNext: const Next
                            }
                            [ text "!!!" ]
         ]
```

#### Pros
- Easy to implement
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Already extensible: developers can define their own attribute, property, output, hook, etc... .
- Supports multiple output handlers.
- Separates type of the data (attributes, properties, etc...).

#### Cons
- Possible to declare an attribute or a property multiple times.

```purs
view :: VirtualNode Message
view = a [ Attr.href "/"
         , Output.click $ const Click
         , Output.blur  $ const Blur
         , Hook.pre     $ const Pre
         ]
         [ Sphenoid.slider [ Attr.step 1
                           , Output.next $ const Next
                           ]
                           [ text "!!!" ]
         ]
```

### PR DESCRIPTION
**Pull request by the-dr-lazy**
I'm trying to implement virtual DOM API as proposed in #5.
Please **review the changes** and help me to make code more readable.

If you have any suggestion about any part of the implementation, please let me know.
If you think any part of implementation is ambiguous and hard to understand, please let me know.
If you think this is the best implementation of virtual DOM API in the universe, please let me know.

### CODE REVIEW

File: example/src/Counter/Main.purs [hunk 1]
```diff
@@ -30,24 +30,21 @@ init = 0
...
+       , button { onClick: const UserClickedIncreaseButton } [ text "+" ]
+       ]
 
 command :: Message
         -> Model
         -> Command (API.Effects ()) Message Output Unit
-command Increase _ = API.increase 
-command Decrease _ = API.decrease 
+command UserClickedIncreaseButton _ = API.increase 
+command UserClickedDecreaseButton _ = API.decrease
```
Comment (the-dr-lazy) [inline example/src/Counter/Main.purs]: Case expression is a better choice for pattern matching in this case.

File: src/Monarch/Type/RowList.purs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import Type.RowList (kind RowList)
+import Type.RowList as RowList
+import Type.Row as Row
+
+class OptionalRecordCons (row :: RowList) (name :: Symbol) (s :: # Type) (t :: # Type)
+
+instance nilOptionalRecordCons :: OptionalRecordCons RowList.Nil _name _s _t
+instance consOptionalRecordCons ::
+  (Row.Union t t' s) => OptionalRecordCons (RowList.Cons name { | t } tail) name s t
```
Comment (the-dr-lazy) [inline src/Monarch/Type/RowList.purs]: `t'` is a trash type variable. It should be marked by underscore.

File: src/Monarch/VirtualDOM.purs [hunk 3]
```diff
@@ -1,40 +1,82 @@
...
+
+type HTMLDivR props hooks message = R HTMLDivElementAttributes (HTMLDivElementOutputs message) props hooks
+
+div :: forall r r' props hooks slots message
+     . Row.Union r r' (HTMLDivR props hooks message)
+    => Row.OptionalRecordCons r "props" (HTMLDivElementProperties ()) props
+    => Row.OptionalRecordCons r "hooks" (Hooks message) hooks
+    => { | r }
+    -> Array (VirtualNode' slots message)
+    -> VirtualNode' slots message
```
Comment (the-dr-lazy) [inline src/Monarch/VirtualDOM.purs]: This type will repeat for other elements. It can become a type synonym!

File: src/Monarch/VirtualDOM.purs [hunk 4]
```diff
@@ -1,40 +1,82 @@
...
+div = h "div"
+
+div_ :: forall slots message. Array (VirtualNode' slots message) -> VirtualNode' slots message
```
Comment (the-dr-lazy) [inline src/Monarch/VirtualDOM.purs]: This type will repeat, too.

File: src/Monarch/VirtualDOM.purs [hunk 5]
```diff
@@ -1,40 +1,82 @@
...
+div_ = h "div" {}
+
+div' :: forall slots message. VirtualNode' slots message
```
Comment (the-dr-lazy) [inline src/Monarch/VirtualDOM.purs]: Also, this one.

**Comment by the-dr-lazy** [comment]
What is your opinion to make properties flat and move attributes under `attrs` field? using of properties is more efficient and predictable so maybe we can encourage using of them in this way.

File: src/Monarch/VirtualDom/NS.purs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module Monarch.VirtualDom.NS where
+
+foreign import kind NS
```
Comment (jedimahdi) [inline src/Monarch/VirtualDom/NS.purs:3]: Where is NS coming from?

File: src/Monarch/Svg.purs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+type Svg = VirtualNode' NS.SVG
+
+-- Elements
+
+type Node r slots message = VirtualDom.Node NS.SVG r slots message
+
+type Node_ slots message = VirtualDom.Node_ NS.SVG slots message
+
+type Leaf r slots message = VirtualDom.Leaf NS.SVG r slots message
```
Comment (jedimahdi) [inline src/Monarch/Svg.purs:29]: What's differences between Node and Leaf? They're defined same.

**Comment by jedimahdi** [comment]
I think nested properties is better than nested attributes. when using a UI library, I don't want specifically say I want to use attributes.
Can you give an example of flatten properties?

File: src/Monarch/VirtualDom/NS.purs (hunk 6 above)
Comment (the-dr-lazy) [reply src/Monarch/VirtualDom/NS.purs:3]: It's PureScript's strange syntax for defining new kinds! Unfortunately PureScript doesn't have DataKinds yet.

**Comment by the-dr-lazy** [comment]
> I think nested properties is better than nested attributes. when using a UI library, I don't want specifically say I want to use attributes.

Let me clarify that I'm arguing about Snabbdom's properties and attributes API.
If I be right every HTML attribute have corresponding DOM property and in fact almost every attribute is just a reflection of a property. So by setting attribute we are just setting a property under the hood. We are in JavaScript, not in HTML so we don't need attribute unless we want to set custom attributes.
Also I doubt the need of attributes API anymore. For custom attributes we should use `dataset` module on Snabbdom.

> Can you give an example of flatten properties?

Using attribute:

```purs
view :: forall message. Html message
view = a { href: "/path", style: "margin: 0" } [ text "Link" ]
```

```js
const a = document.createElement('a')

a.setAttribute('href', '/path') // this will changes href property.
a.setAttribute('style', 'margin: 0') // should be parsed and also changes the style property.
```

Using property:

```purs
view :: forall message. Html message
view = a { href: "/path", style: { margin: '0' } } [ text "Link" ]
```

```js
const a = document.createElement('a')

a.href = '/path'
a.style.margin = '0'
```

**Comment by the-dr-lazy** [comment]
I'm going to close this PR to make it unit with #11.

**Pull request closed without merging**